<html><head>	<title>TiMAT2 (T2) Applications</title><style type="text/css">	#cr{text-align:center; color: #000000; font-weight: bold}	#rt{text-align:right; color: #000000; font-weight: bold}	#grBk {background-color: #CC9966;}	TD {font-family: Verdana, Arial, Helvetica, sans-serif; font-size:12;}	H1 {color: #996633; font:arial; font-size:16;}	H2 {color: #996633; font:arial; font-size:12;}	BODY {color:black; background-color:white; font-family: Verdana, Arial, Helvetica, sans-serif; font-size:12;}		A:link    {text-decoration: none; color: #000000; font-weight: bold}  	A:visited {text-decoration: none; color: #000000; font-weight: bold}   	A:hover   {text-decoration: none; color: #FFCC66; font-weight: bold} 	A:active  {text-decoration: none; color: #000000; font-weight: bold}   </style></head><body><center><H1>Core Applications</H1>Launch these applications using the -jar option, for example, 'java -jar pathTo/T2/Apps/FileSplitter' .<br>Click the name to see the command line menu.<p>		<table ID='grBk' width="750" cellspacing="5" cellpadding="2" >	<tr><td ID='cr' COLSPAN="2">Cel File Apps</tr><tr><td ID='rt'><a href="cmdLnMenus.html#CelFileQualityControl">Cel File Quality Control</a><td>Calculates various QC statistics on groups of xxx.cela files, including hierarchical clustering. Renders and saves png virtual slides for outlier files.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#CoordinateExtractor1lq">Coordinate Extractor 1lq</a><td>Extracts control and pm coordinates from a text 1lq file for use by CelFileQualityControl app. Also creates virtual cel files for each class for visualization using the VirtualCel app. Useful for confirming correct orientation.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#CelFileConverter">Cel File Converter</a><td>Converts text cel files into serialized float[][] arrays (xxx.cela) for use by other TiMAT2 applications.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#HierarchicalClustering">Hierarchical Clustering</a><td>Visually clusters serialized float[] (xxx.celp) arrays using a Pearson correlation coefficient. Good for QC.</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#Correlate">Correlate</a><td>Outputs all pairwise correlations between serialized float[] (xxx.celp) arrays using a Pearson correlation coefficient as well as performing heirarchical clusters.</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#CelProcessor">Cel Processor</a><td>Walks through the TPMap object file fetching the PMMM intensity values, saves the float array to disk. Provides options for median scaling, quantile normalization and PM-MM transformation.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#VirtualCel">Virtual Cel</a><td>Creates image files for each text version '.cel' file provided.  Useful for identifying chips in need of masking by the Cel Masker application.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#CelMasker">Cel Masker</a><td>Draws a virtual cel file base on a text version 'xxx.cel' file.  Problem areas can be circled, masked, and saved as text version 'xxx.cel' files.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntensityPrinter">Intensity Printer</a><td>Prints .sgr files for any serialized int array (i.e. the processed cel files, xxx.cela).	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ConvertAgilentData">Convert Agilent Data</a><td>Parses Agilent two color tiling data files to tpmaps and cela files for processing in TiMAT2.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ConvertGeoData">Convert Geo Data</a><td>Parses two color Gene Expression Omnibus data files to tpmaps and cela files.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ConvertNimblegenPAIR2Cela">Convert Nimblegen PAIR 2 Cela</a><td>Converts Nimblegen PAIR text data file(s) to cela files for processing in TiMAT2.	</tr><tr><td ID='cr' COLSPAN="2">&nbsp;</tr><tr><td ID='cr' COLSPAN="2">TPMap Apps</tr><tr><td ID='rt'><a href="cmdLnMenus.html#MummerMapper">Mummer Mapper</a><td>Uses <a href="http://sourceforge.net/projects/mummer">Mummer (http://sourceforge.net/projects/mummer)</a> to map and filter a 1lq text file against a genome.  100x faster than the TPMap Oligo Blast Filter and recommended if you don't care about 1bp mismatches. For processing very large chromosomes, ie human chr2 or 3 but not any drosophila, mummer needs a 64 bit machine with lots of RAM.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#TPMapOligoBlastFilter">TPMap Oligo Blast Filter</a><td>Blasts each oligo against a genome. Counts the number of matches (exact matches and 1bp mismatches). Replaces the t/f orientation column in the TPMap 												file with this count.  Reassigns the chromosomal coordinates. Writes two files: BlastScored, saves TPMap lines that have at least one exact match; BlastFiltered, saves only those TPMap lines that have one 												exact match. Provides statistics on the filtering.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#TPMapProcessor">TPMap Processor</a><td>Converts a tpmap file into precomuted windows across each chromosome for use by other TiMAT2 applications.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#TPMapSort">TPMap Sort</a><td>Sorts a TPMap by chromosome and start position.  Use to sort the remapped TPMap file.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#FilterTPMapByRegions">Filter TPMap By Regions</a><td>Strips any tpmap oligos not contained within one or more of the user defined regions.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ConvertNimblegenNDF2TPMap">Convert Nimblegen NDF 2 TPMap</a><td>Converts a Nimblegen NDF text file to a tpmap.</tr><tr><td ID='cr' COLSPAN="2">&nbsp;</tr><tr><td ID='cr' COLSPAN="2">Main T2 Apps</tr><tr><td ID='rt'><a href="cmdLnMenus.html#T2">T2</a><td>Launches many of the TiMAT2 applications based on a tab delimited parameter file to convert, normalize, split, and merge txt cel files before launching ScanChromosomes.  Runs on one machine or in parallel on a cluster.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#FDRWindowConverter">FDR Window Converter</a><td>Associates an empirical FDR esTiMAT2e with each window provided mock IPs were performed.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#FindSubBindingRegions">Find Sub Binding Regions</a><td>Finds the highest average scoring intensity difference or ratio sub window, within an interval, typically 350bp.  Also automatically picks binding peaks and binding regions.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntervalFilter">Interval Filter</a><td>Filters intervals by a variety of criteria into two files, pass or fail.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntervalGFFPrinter">Interval GFF Printer</a><td>Prints a text GFF3 file to represent the Intervals.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntervalGraphPrinter">Interval Graph Printer</a><td>Prints a .sgr file to represent the Intervals in Affymetrix's IGB.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntervalMaker">Interval Maker</a><td>Merges overlapping windows into intervals.  Options are provided to set the minimum required overlap and minimum score needed for merging.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntervalPlotter">Interval Plotter</a><td>Displays a detailed picture for each interval with oligo intensity information, transcription factor binding sites, repeat regions, best windows, etc.  Use as the final filter in deciding whether the Interval is reasonable.  Also use to pick binding regions and binding peaks by clicking and dragging.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntervalReportPrinter">Interval Report Printer</a><td>Generates a spreadsheet or detailed reports for the Intervals with an assortment of calculations.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#LoadChipSetIntervalOligoInfo">Load Chip Set Interval Oligo Info</a><td>Similar to Load Interval Oligo Info but works on data from multi chip set experiments. </tr> <tr><td ID='rt'><a href="cmdLnMenus.html#LoadIntervalOligoInfo">Load Interval Oligo Info</a><td>Fetches and saves oligo intensity information from processed cel files for each interval.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#MakeChromosomeSets">Make Chromosome Sets</a><td>For use with multi chip sets. MakeChromosomeSets takes directories of chip set directories containing split chromosome intensity files from the CelProcessor app and combines them into a master set of chromosome split intensity files for use by ScanChromosomes. </tr><tr><td ID='rt'><a href="cmdLnMenus.html#MergeWindowArrays">Merge Window Arrays</a><td>Concatenates serialized Window[]s into one. Useful for combining chromosome specific Window array results from the Scan Chromosome application.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#MultiWindowIntervalMaker">Multi Window Interval Maker</a><td>Merges overlapping windows into intervals from multiple Window[] arrays/ multiple experiments. Options are provided to set the minimum required overlap and minimum score needed for each Window[] array for merging, as well as the minimum number of passing windows, ie 2 out of 3.  Useful for combining replicas or different antibodies into composit Intervals when one doesn't want to process them as a pool.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#OligoIntensityPrinter">Oligo Intensity Printer</a><td>Prints .sgr files for the intensity ratio and intensity difference for each oligo when given processed treatment and control cel files.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#OverlapCounter">Overlap Counter</a><td>Sorts two sets of Intervals based on whether they overlap one another.  Good for finding common sets of Intervals between say two different antibodies or excluding intervals found associated with controls.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ScanChip">Scan Chip</a><td>Use to score windows of oligo intensities with a Wilcoxon Rank Sum test and a trimmed mean or pseudo median ratio test.  Wraps Richard Bourgon's symmetric p-value test and John Storey's q-value multiple testing applications to convert window ratio scores into corrected p-values. Saves a variety of window and point sgr files for direct visualization to Affymetrix's IGB.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ScanChromosomes">Scan Chromosomes</a><td>Similar to Scan Chip but works on chromosomes.  Useful for processing multi chip sets, ie those where the genome is spread out over several chips, not just one.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ScanGenes">Scan Genes</a><td>Extracts treatment and control intensities for all exons in each gene.  Calculates a simple median ratio and ranks the genes by expression.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ScoreIntervals">Score Intervals</a><td>Scores Intervals for hits to a transcription factor binding matrix, LLPSPM.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#SetNumberIntervalMaker">Set Number Interval Maker</a><td>Determines the threshold(s) needed to generate a set number of intervals, multiple score indexes accepted. Useful for stratifying your results into say top 125, 250, 500 ... regions.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#WindowScanner">Window Scanner</a><td>Counts the number of windows that pass a range of score cutoffs. Use this info to calculate false discovery rates.</tr><tr><td ID='cr' COLSPAN="2">&nbsp;</tr><tr><td ID='cr' COLSPAN="2">General Analysis Apps</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#AggregatePlotter">Aggregate Plotter</a><td>Fetches point data contained within each region, zeros the coordinates, scales, sums, and window averages the values.  Usefull for generating class averages from a list of annotated regions. Use a spreadsheet app to graph the results.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#FindNeighboringGenes">Find Neighboring Genes</a><td>FNG takes a list of genes in UCSC Gene Table format and intersects them with a list of regions finding the closest gene to each region as well as all of the genes that fall within a given neighborhood. </tr><tr><td ID='rt'><a href="cmdLnMenus.html#AnnotateRegions">Annotate Regions</a><td>Use to characterize what genes surround the binding regions and whether there is a bias in their distribution.  Can be combine with programs like GOMiner to identify possible biological functions associated with  ChIP Chip data.  At present it works with Drosophila 4.0 annotation.  Parsers for other gff annotation can be readily adapted.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntersectRegions">Intersect Regions</a><td>Performs an intersection analysis on lists of genomic regions, uses random regions matched for GC content, length, array interrogated regions, and chromosome to calculate an enrichment over random and p-value. Also generates a distance to nearest region distribution histogram.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#IntersectLists">Intersect Lists</a><td>Intersects two lists (of gene names) and using randomization, calculates the significance of the intersection and the fold enrichment over random.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ParseBarsForParticularRegions">Parse Bars For Particular Regions</a><td>Given a list of regions and a directory of graph files, will collect and print the associated scores.  A p-value and fold enrichment for each region's score can be calculated using chromosome, length, # scores, and gc matched random regions.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#RankedSetAnalysis">Ranked Set Analysis</a><td>Performs an intersection analysis on lists of ranked regions creating a visual box-line-box representation as well as a rank based % intersection graph.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#CorrelationMaps">Correlation Maps</a><td>CM creates correlation maps from gene expression data to look for physical gene clusters (aka gene expression neighborhoods, chromosome territories).</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#ScatterPlot">Scatter Plot</a><td>Very simple scatter plot and correlation coefficient calculator for huge int or float arrays.  Use to estimate data reproducibility. For example, the correlation coefficient for two independent ChIP Chip experiments using the same antibody should be above 0.8 and preferably above 0.9.  If not, optimization is needed.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ScoreChromosomes">Score Chromosomes</a><td>Scores a genome for hits to a transcription factor binding matrix, LLPSPM.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ScoreSequences">Score Sequences</a><td>Scores a multi-FASTA file of sequences for hits to a transcription factor binding matrix, LLPSPM.	</tr><tr><td ID='cr' COLSPAN="2">&nbsp;</tr><tr><td ID='cr' COLSPAN="2">Utility Apps</tr><tr><td ID='rt'><a href="cmdLnMenus.html#JQSub">JQSub</a><td>JQSub executes a given command line on the specified clusters via qsub.  It provides extensive logging and tracking of your submission.</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#Primer3Wrapper">Primer 3 Wrapper</a><td>Wrapper for the <a href="http://frodo.wi.mit.edu/primer3/">primer3 (http://frodo.wi.mit.edu/primer3/)</a> application. Extracts sequence, formats for primer3, executes, and parses the output to a spreadsheet. Useful for bulk qPCR primer picking. Yes, you do need to validate your results.</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#Bar2Gr">Bar 2 Gr</a><td>Converts xxx.bar files to text xxx.gr files.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#Gr2Bar">Gr 2 Bar</a><td>Converts xxx.gr.zip files to binary chromosome specific xxx.bar files.  Very fast loading in IGB.  Small size.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#Sgr2Bar">Sgr 2 Bar</a><td>Converts xxx.sgr.zip files to binary chromosome specific xxx.bar files.  Very fast loading in IGB.  Small size.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#Tag2Bar">Tag 2 Bar</a><td>Converts signature sequencing xxx.bed files (tab delimited chr start stop etc) to binary chromosome specific stair-step/ heatmap xxx.bar files.  Very fast loading in IGB.  Small size.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#Windows2HeatMapSgr">Windows 2 Heat Map Sgr</a><td>Converts a text file containing four tab delimited columns (chr, start, stop, score) to a heat map compatible file for import into IGB.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ExportIntergenicRegions">Export Intergenic Regions</a><td>Takes a GFF file and exports regions not covered by any annotation, the intergenic regions.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#ExportIntronicRegions">Export Intronic Regions</a><td>Takes a UCSC gene table and exports the most conservative estimate of intronic sequence.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#BestWindowScoreExtractor">Best Window Score Extractor</a><td>Fetches the best window for each given region from a serialized Window array.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#FetchGenomicSequences">Fetch Genomic Sequences</a><td>Given a file containing genomic coordinates, fetches and saves the sequence (column output: chrom origStart origStop fetchedStart fetchedStop completeFetch seq).</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#OligoTiler">Oligo Tiler</a><td>Tiles oligos across genomic regions returning their forward and reverse sequences.</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#MergeRegions">Merge Regions</a><td>Merges genomic regions. Good for collapsing duplicate regions for creating microarray seq capture designs.</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#SubtractRegions">Subtract Regions</a><td>Removes regions from a genomic regions file. Good for RepeatMasking genomic regions prior to tiling.</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#RandomizeTextFile">Randomize Text File</a><td>Randomizes the lines in a text file.  Good for randomly selecting reverse oligos to pad a microarray design.</tr>	<tr><td ID='rt'><a href="cmdLnMenus.html#FileCrossFilter">File Cross Filter</a><td>FCF take a column in the matcher file and uses it to parse the rows from other files. Useful for pulling out and printing in order the rows that match the first file.</tr><tr><td ID='rt'><a href="cmdLnMenus.html#FileJoiner">File Joiner</a><td>Joins many text files together, paying attention to avoid fusing the last and first lines from two files.  Use to combine the remapped TPMap file.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#FileSplitter">File Splitter</a><td>Splits a text file into many files given a number of lines, use to split the TPMap file for TPMapOligoBlastFiltering.	</tr><tr><td ID='rt'><a href="cmdLnMenus.html#PrintSelectColumns">Print Select Columns</a><td>Spread sheet/ tab delimited file manipulations.</tr></center></table>	</body></html>