<html><head>	<title>Usage</title><style type="text/css">	#rt{text-align:right; color: #000000; font-weight: bold}	#grBk {background-color: #CC9966;}	TD {font-family: Verdana, Arial, Helvetica, sans-serif; font-size:12;}	H1 {color: #996633; font:arial; font-size:14;}	H2 {color: #996633; font:arial; font-size:12;}	BODY {color:black; background-color:white; font-family: Verdana, Arial, Helvetica, sans-serif; font-size:12;}		A:link    {text-decoration: none; color: #000000; font-weight: bold}  	A:visited {text-decoration: none; color: #000000; font-weight: bold}   	A:hover   {text-decoration: none; color: #FFCC66; font-weight: bold} 	A:active  {text-decoration: none; color: #000000; font-weight: bold}   </style></head><H1>TiMAT2 Usage</H1><br><body>This is an outline on how to use TiMAT2 (aka T2) to process your tiling microarray data.  <p>The applications are designed to be user friendly and require only moderate familiarity with command line programs.  When in doubt type 'java -jar pathTo/T2/Apps/ApplicationName' to pull a menu of options.See the <a href="README.html">readme file</a> for installation instructions.<p><h1>General Recommendations:</h1><ol><li> I think it important to have three replicas for every experiment.  These don't necessarily have to be biological replicas, e.g. three different isolates of a fibroblast cell line processed in parallel.  Technical replicas in most cases are preferred, e.g. pool multiple plates of fibroblasts and split it after fixing/lysing/preparing chromatin and process in parallel.  Technical replicas are extremely informative.  They provide a measure of the sample prep/ microarray processing consistency without any confounding biological variability.  </li><li> If you choose to only go with two replicas be prepared for lots of qPCR validation.  The symmetric null method of confidence estimation works reasonably well for non complex organisms but rarely for human or mouse.  Three IPs and three input replicas are the minimum needed for a random label permutation confidence estimation. </li><li> It is rather critical to perform all of these preps using the same reagents, same equipment, at the same time, etc. Be prepared to throw out all of the data generated during the pilot/ optimization phase of the experiment and repeat it along side the other samples.  If you do plan to make use of the pilot data, prep/ IP/ amplify in parallel the other technical replicas and freeze them while deciding if the pilot worked, if so, then run the matched samples.  One good way to kill a tiling experiment is to prep samples on different days using different reagents and expect some sort of magic to happen upon hybridization and anlysis. Technical inconsistency is the primary cause for tiling microarray failure.</li><li> For two color arrays, there is no need to have a matching input control on every chip.  You can treat the different colors independently and think of them as a way of getting two readouts from the same chip.  For example, if I performed a ChIp-ChiP experiment with 4 antibodies on the same fibroblast cells I would use 9 chips. For each antibody, I would prepare three technical replicas, and label them with Cy3 or Cy5 and place them on different chips.  I would do likewise with one input sample split to three technical replicas. For the last sample, I would perform mock IPs, three technical replicas, using IgG or a non specific antibody, e.g. anti-GST or anti-FLAG. See below.</li><li> For your experiments involving an IP, consider performing a mock IP using IgG or an antibody that doesn't bind anything within your sample, e.g. anti-GST or anti-FLAG.  It is best to use a type matched antibody.  TiMAT2 is set up to generate empirical FDRs based on the mock IP and is a great way to estimate confidence in your real IP data without having to perform exhaustive qPCR. One consideration though, it must be performed in parallel with the real IPs.  Subtle differences in washing and amplification really effect the output.</li> <li> Don't over wash your beads during the IP.  Use the minimum number of washes (3x 5min?) that give good fold enrichment for known targets by qPCR.  Over washing can lead to a PCR bottle neck effect where a small number of regions get overly amplified and will look exactly like real regions on the microarray even though nothing specific was pulled down by the IP. </li><li> You might want to prepare twice the amount of chromatin you will need to be sure to not run out if and when parts of the experiment need to be repeated.</li><li> Save enough pooled pre amplification real IP and input material for ~50 qPCR rxns each.</li></ol><h2>For ChIP-chip experiments:</h2><ol><li> Perform:	<ul>		<li> Three technical replicas for each anti-transcription factor antibody (i.e. hybridize three independent parallel processed IPs to three chips)		<li> Three technical replicas for a mock IgG IP.		<li> Three technical replicas for just input chromatin.		<li> Commercial chips are very consistent. Technical repeats (using the same hybridization solution on different chips) are unnecessary and can emphasize false positives in certain testing situations.	</ul><li> For Affy data, skip mismatch transformation for ChIP-Chip experiments. Just median scale intensity values to 100 and quantile normalize across all chips.<li> Use the <a href="cmdLnMenus.html#Correlate">Correlate</a> app to check the correlation coefficient between your replicas and appropriate clustering of like chips.For IPs, the r^2 values between replicas should be > 0.5 (or 50 for 100 *r^2). These should typically be > 0.8. If they are less than these values your IP/ PCR/ labeling protocols must be optimized. It cannot be emphasized enough to not proceed with the full experiment until thesevalues have been achieved. Otherwise, considerable time and money will be wasted.  Garbage in = garbage out. Speckles and spots on the chip can throw the correlation coefficient.  If this is the case, use full quantile normalization and the robust pseudo median window scanning statistic. </ol><br><h1>Required Resources:</h1><ol><li> Java version 1.4 or greater. <li> These command line applications have been tested on MacOSX and Linux.  They have not been tested on a Windows machine. All the source code, with extensive documentation, is included in the TiMAT2 package. <li> A computer with > 2 GigaBytes RAM.  One can use a machine with less memory but some programs will run rather slowly, hours instead of minutes. </ol><br><h1>Processing Protocol:</h1><h2>Tiling analysis follows five basic steps:</h2><ol><li> Obtain or make map file(s) to associate microarray intensities with genomic coordinates, release specific<li> Normalize raw microarray intensities<li> Map normalized intensities to the genome<li> Use window smoothing to create an array of window summary scores<li> Identify enriched (and or reduced) regions by picking a threshold summary score and merge overlapping windows that exceed the threshold</ol><br><h2>Obtaining Map files</h2><ol><li> TiMAT2 uses the Affymetrix tpmap file format.  This is a text version of the binary probe map file, bpmaps.  A bpmap can be converted to a tpmap using the bpmap2tpmap command line utility.  Since this tool is not officially released ask your Affy rep for the tpmap version of the officially released bpmaps.<li> TPMaps can be generated from the NimbleGen NAD files using the <a href="cmdLnMenus.html#ConvertNimblegenNDF2TPMap">ConvertNimblegenNDF2TPMap</a> app.<li> TPMaps might be generated from the Agilent txt results file, provided sequences are present, using the <a href="cmdLnMenus.html#ConvertAgilentData">ConvertAgilentData</a> app.<li> You can create your own, it's a simple tab delimited txt file containing:	<ol>		<li> Two comment lines: '#seq_group_name' and '#version'		<li> Oligo sequence (5'->3', sense strand)		<li> t or f (boolean describing whether the real oligo on the microarray will hybridize with the sense target strand)		<li> Chromosome		<li> Start position of the oligo, zero base coordinates		<li> Perfect match X microarray coordinate		<li> Perfect match Y microarray coordinate		<li> (Optional) Mis match X microarray coordinate		<li> (Optional) Mis match Y microarray coordinate		<li> (Optional) Mystery number 1 (for backward compatibility?)	</ol><li> Example of a Sense, PM only tpmap<pre>#seq_group_name#versionACAATGTACAATACCGAAGGACTATTGCAGATAAATTCTTACAAGGTGATGTACGTGATC    t       chr1    0       290     2061TGTAGGTCATGGGCTAGAAAGGCAAACTTGATATCGTTAGTTTACTATGCTTGGAACAAT    t       chr1    55      489     2571TTGTGGTTTGACAAAGTTTGAACATGCCGACAGTGATATGGCTTACACTTCTGACTGTAG    t       chr1    110     422     781ACCACAACTTGCGAGAATGTAAACTACGTATTCGATTTCAAGGACTTTGTTTATTTTGTG    t       chr1    165     108     2431ACCGGGAAAGCATGCCCCGTACGCTTATCTACTTGTTATAAATGTGGCAAGGCCGACCAC    t       chr1    220     25      3181AGATGAATGTCTAAGTTGCACTCTTGGTAACCCAGACACAGAAATTCGTGCTCATACCGG    t       chr1    275     279     411GGGTAGACTCAACTGACATTTCAGCACAACTCGAGCGATTCTTTCGAGTGTATAAAGATG    t       chr1    330     130     371CGTACCATTCATTTCGCCGAGTATAAAAGTCGTGTGCAAGCCGTCAAAAAACAATGGGTA    t       chr1    385     519     951...truncated...</pre></ol><br><h2>Building Map Files</h2>Building map files is a very complex process fraught with many pitfalls.  It should only be performed by tiling microarray professionals.<ol><li> Obtain fasta files for the genome you wish to map. Complex organisms should be RepeatMasked with X's or N's, not lower case.<li> Convert the fasta headers into the convention used by UCSC and IGB. (e.g. > chromosome 1 : NCBI build 35.1 to >chr1)<li> (Optional) Obtain control sequence fasta files containing control sequences (bacterial, arabidopsis, intergenic regions...) include these with the genomic sequences.  You may also like to create one particular fasta file and call it chrCtrls.fasta.  This can be used in median scaling during array normalization. <li> Create or obtain a xxx.1lq file.  This can be obtained directly from your Affy rep. Do not use a rotated 7G specific xxx.1lq file.  For NimbleGen and Agilent, you will need to create it. The <a href="cmdLnMenus.html#PrintSelectColumns">PrintSelectColumns</a> app may be helpful.  You'll need to copy and paste entire columns and most spread sheet applications cannot handle the large sized map files. You may also need to transform the array coordinates into zero based coordinates. It is a tab delimited txt file containing:	<ol>		<li> A variety of header lines only one is required 'X	Y       Seq     Destype'		<li> X coordinate on the array (zero based)		<li> Y coordinate on the array (zero based)		<li> Oligo sequence (the original 1lq format specified 3'-5', you can leave it 5'-3' and throw the appropriate flag in <a href="cmdLnMenus.html#MummerMapper">MummerMapper</a>)		<li> Destype (the original 1lq format specified a number used to describe the class of oligo, PM, MM, S, AS, Bright Control, Dark Control, etc. You can put anything here and use a regular expression in <a href="cmdLnMenus.html#MummerMapper">MummerMapper</a> to select appropriate probes to map.)	</ol><li> Example of a custom xxx.1lq file that can be read by <a href="cmdLnMenus.html#MummerMapper">MummerMapper</a>: <pre>X       Y       Seq     Destype0       3       TATATTTCTGCATATACATCATAGTCTCTAAAAACTGTACTAGGTCATACTCCATAGAGG    Spo_P101594-R|Spombe|complement_chr2:8636..86950       4       AGTTATGACGAGCTTTGGAACGTTTATGACGACTGTTAGGATCGCAGTTTCCTGAAGTCT    Spo_P110716-R|Spombe|complement_chr2:511391..5114500       5       TCGTAGACCATCAAAAACTGTAACGAATAGTTCCAAAACGGTCACCACTAGACTTAGCAC    Spo_P022386|Spombe|chr1:1232221..12322800       6       GACAATAATTTGGCGGCGCTGCTACCAATATAGCTGGAAGCAGTAAAGTATTAATCGTGC    Spo_P043768-R|Spombe|complement_chr1:2408231..24082900       7       TAGGCTAAGCATCGAGTTCATGGCTAAATTAATAAAAGACTTTAAAAGTTTGTTAAAATG    Spo_P000762-R|Spombe|complement_chr1:42901..429600       8       CCACGACCTTTTCCGTGAGTCCGGGGGTTGTAAGTCTTCTTTATACCGTAGCGGGTAGAA    Spo_P041016-R|Spombe|complement_chr1:2256871..22569300       9       CATCAGCTATGAACTTTCACTTTTGTCTACAACTTTTTTAATCAGCTACAAATTTACAAA    Spo_P042235-R|Spombe|complement_chr1:2323916..23239750       10      CAGGATTACAACAGGTTAAATGTCGATTTAACAAGGAAAGACTTACGTCACCACAGAGAA    Spo_P105785|Spombe|chr2:240186..240245...truncated...</pre><li> Run the <a href="cmdLnMenus.html#MummerMapper">MummerMapper</a> app.  Questions to address:	<ul>		<li> How many exact matches to allow?		<li> Stranded? S, AS, or Both?		<li> Which probe Destypes?		<li> (Affy specific) Is this to be used on a pre 7G or 7G scanner?		<li> Are you loading large mammalian chromosomes?  Run on a 64 bit machine!		<li> Do you want partial matches? What is the minimum sized oligo to match?		<li> Do you need to reverse the oligo sequences?		<li> Are MM oligos present?	</ul><li> Test, test, and test the finished tpmap using known datasets. One mistake may create hundreds of hours of analysis headaches later on.  Building tpmaps is a lot like taking pictures for your friend's wedding.  Sounds easy but mess it up and you'll be mistrusted for life.</ol><br><h2>Convert Raw Microarray Data Into xxx.cela Files</h2>To use T2 you must convert your raw microarray intensity data into a format that can be read by TiMAT, serialized binary Java float[][] files.  These are compressed, load quickly, and are used by multiple TiMAT2 applications.<ul><li> Affymetrix text xxx.cel files can be converted to xxx.cela using the <a href="cmdLnMenus.html#CelFileConverter">CelFileConverter</a> app.<li> Affymetrix binary xxx.cel files can be converted to text xxx.cel files using the <a href="http://www.affymetrix.com/Auth/support/developer/downloads/Tools/CelFileConversion.ZIP">CelFileConversion tool</A><li> Agilent xxx.txt results files are converted to xxx.cela using the <a href="cmdLnMenus.html#ConvertAgilentData">ConvertAgilentData</a> app.  This application parses the processed intensities and raw intensities.  The later are recommended for use in T2.<li> NimbleGen text xxx.PAIR data files are converted to xxx.cela using the <a href="cmdLnMenus.html#ConvertNimblegenPAIR2Cela">ConvertNimblegenPAIR2Cela</a> app.</ul><br><h2>Option 1: Process your tiling data using the T2 application (Standard Analysis)</h2>The <a href="cmdLnMenus.html#T2">T2</a> application is a wrapper for many TiMAT2 applications and can work with a cluster to farm out big jobs for rapid analysis.<ol><li> Convert your raw intensity data to binary xxx.cela or text xxx.cel files, see above.<li> Create processed TPMapFiles by running the <a href="cmdLnMenus.html#TPMapProcessor">TPMapProcessor</a> app on your tpmap file(s).<li> Complete the <A href="t2ParamFileTemplate.txt">t2ParamaterFile</a> and launch the <a href="cmdLnMenus.html#T2">T2</a> app.</ol> <br> <h2>Option 2a: Process your <i>single</i> chip set tiling data using individual TiMAT applications (Custom Analysis)</h2>This approach can be used to customize your analysis for non standard tiling experiments. Many options exist here that are not available in the T2 app.<ol><li> Convert your raw intensity data to binary xxx.cela files, see above.<li> Create processed TPMapFiles by running the <a href="cmdLnMenus.html#TPMapProcessor">TPMapProcessor</a> app on your tpmap file(s)., pay close attention to the size of the window option.<li> (Optional) Run the <a href="cmdLnMenus.html#VirtualCel">VirtualCel</a> application on a directory of converted cel files (xxx.cela) to create image files for visual inspection. Any large spots should be masked using the <a href="cmdLnMenus.html#CelMasker">CelMasker</a> app.<li> Run the <a href="cmdLnMenus.html#CelProcessor">CelProcessor</a> app on sets of xxx.cela files to scale, normalize, and map the data.  This application generates xxx.celp files.<li> (Optional) Use the <a href="cmdLnMenus.html#ScatterPlot">ScatterPlot</a> to draw a simple scatter plot and calculate a Pearson correlation coefficient on processed cel files. For example, compare the different treatment chips to one another. There should be a good internal correlation (>=0.8). <li> (Optional) Use <a href="cmdLnMenus.html#HierarchicalClustering">HierarchicalClustering</a> to cluster your xxx.celp files.  Any file that fails to cluster with an r^2*100 value of > 50 should be removed from the analysis.<li> Run <a href="cmdLnMenus.html#ScanChip">ScanChip</a><li> (Optional) Run <a href="cmdLnMenus.html#ScanGenes">ScanGenes</a> to perform a basic expression array analysis using your tiling data.<li> If you have performed mock IPs, run the <a href="cmdLnMenus.html#FDRWindowConverter">FDRWindowConverter</a> app to calculate an empirical FDR for each window.<li> Merge high scoring, overlapping Windows into an array of Intervals with the <a href="cmdLnMenus.html#IntervalMaker">IntervalMaker</a> application. The biggest difficulty is where to set the threshold for merging windows. Two confidence estimations are provided by TiMAT2: an empirical FDR based on a mock IP and a statistical FDR based on Richard Bourgon's non-parametric symmetric p-test. Set these generously and manually filter. If you have multiple replicas, or have used different antibodies you can merge the different Window arrays using the <a href="cmdLnMenus.html#MultiWindowIntervalMaker">MultiWindowIntervalMaker</a> app.<li> (Optional) The <a href="cmdLnMenus.html#SetNumberIntervalMaker">SetNumberIntervalMaker</a> can also be used to make multiple interval arrays, each containing a specific number of intervals.  This is useful, and recommended, for analysis independent of thresholds.<li> Load the interval array(s) with oligo information using the <a href="cmdLnMenus.html#LoadIntervalOligoInfo">LoadIntervalOligoInfo</a> app.<li> For chIP-chip data, find the best average intensity difference sub window within each Interval, as well as enrichment peaks using <a href="cmdLnMenus.html#FindSubBindingRegions">FindSubBindingRegions</a>.<li> (Optional) Score Intervals for the presence of a transcription factor using <a href="cmdLnMenus.html#ScoreIntervals">ScoreIntervals</a>. You can also score whole chromosomes or a set of FASTA sequences using <a href="cmdLnMenus.html#ScoreChromosomes">ScoreChromosomes</a> and <a href="cmdLnMenus.html#ScoreSequences">ScoreSequences</a> respectively.<li> (Optional) Filter Intervals with a variety of parameters using <a href="cmdLnMenus.html#IntervalFilter">IntervalFilter</a>. Sorts intervals into pass and fail.<li> (Optional) Compare and split different Interval sets based on their overlap with the <a href="cmdLnMenus.html#OverlapCounter">OverlapCounter</a>.  Use the <a href="cmdLnMenus.html#IntersectRegions">IntersectRegions</a> app for a more robust and sophisticated analysis.<li> Print results:	<ul>		<li> Print interval plots for processed intervals using <a href="cmdLnMenus.html#IntervalPlotter">IntervalPlotter</a>. These are graphic representations of the individual oligo intensities for each data set, the averaged treatment chip, the averaged control chip, the intensity difference, the intensity ratio, smoothed (trimmed mean) ratios, as well as hits to a position specific probability matrix, number of matches, and the best window, and the best sub window. These plots can be saved to disk as PNG files or manipulated directly to print chromosomal coordinates for a point or region (including the genomic sequence).		<li> Print interval reports in a spread sheet or page format using <a href="cmdLnMenus.html#IntervalReportPrinter">IntervalReportPrinter</a>.		<li> Print Intervals as .sgr files for import into Affymetrix's IGB browser with <a href="cmdLnMenus.html#IntervalGraphPrinter">IntervalGraphPrinter</a>.		<li> Print all the oligo values as .sgr files for import into IGB using <a href="cmdLnMenus.html#OligoIntensityPrinter">OligoIntensityPrinter</a>.		<li> Print any serialize int[] (ie processed xxx.celp files) associated with each oligo in the tpmap file as .sgr files for import into IGB using <a href="cmdLnMenus.html#IntensityPrinter">IntensityPrinter</a>.		<li> Print the Intervals in GFF3 format with <a href="cmdLnMenus.html#IntervalGFFPrinter">IntervalGFFPrinter</a>.	</ul></ol><br><h2>Option 2b: Process your <i>multi</i> chip set tiling data using individual TiMAT applications (Custom Analysis)</h2>Multi-chip set analysis is very similar to single-chip set analysis and is the pipeline wrapped by the T2 app. The key differences are noted below.<ol><li> Set the -r flag when running the <a href="cmdLnMenus.html#CelProcessor">CelProcessor</a> app to break the mapped intensities down by chromosome.  The results in this case will be a directory instead of a single xxx.celp file for each array.<li> Run the <a href="cmdLnMenus.html#MakeChromosomeSets">MakeChromosomeSets</a> app to merge the chromosome specific normalized data directories together.<li> Run <a href="cmdLnMenus.html#ScanChromosomes">ScanChromosomes</a> instead of <a href="cmdLnMenus.html#ScanChip">ScanChip</a>.<li> Run <a href="cmdLnMenus.html#MergeWindowArrays">MergeWindowArrays</a> if you processed individual chromosomes in <a href="cmdLnMenus.html#ScanChromosomes">ScanChromosomes</a>.<li> Use the <a href="cmdLnMenus.html#LoadChipSetIntervalOligoInfo">LoadChipSetIntervalOligoInfo</a> instead of the <a href="cmdLnMenus.html#LoadIntervalOligoInfo">LoadIntervalOligoInfo</a>.</ol><br><h1>Useful higher level analysis and utility applications</h1>Be sure to check out the <A href="applications.html">General Analysis and Utility applications</A>.  They might save you days of coding.</body></html>